core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -373,6 +373,11 @@ private void simplifyOrTerms(List<RexNode> terms) {
 
   private RexNode simplifyNot(RexCall call) {
     final RexNode a = call.getOperands().get(0);
+    if (a.isAlwaysTrue()) {
+      return rexBuilder.makeLiteral(false);
+    } else if (a.isAlwaysFalse()) {
```

Comment:
Please avoid use of `isAlways...` here. Please optimize literals only.

Consider `not(not(not(not(vBool()` case. Then `isAlways..` would go deep into the tree and return false.
It would result in `O(N^2)` behavior which we don't want in `RexSimplify`.